### ISSUE TITLE
/table/ paths with 500 errors - C tables do not exist

### ISSUE THREAD

**Issue by tuchandra** [censusreporter/censusreporter#176]
From issue #162, we identified 4 categories of /table/ paths that return 500 errors. This is category 4 of 4.

These tables do not exist, but the corresponding B tables do. In all cases but `C12002`, there is a topics page that lists this tabulation as having a collapsed version (Migration for `C07203`, Commute for `C08017` and `C08602`). 
http://censusreporter.org/tables/C07203/
http://censusreporter.org/tables/C08017/
http://censusreporter.org/tables/C08602/
http://censusreporter.org/tables/C12002/

Resolution of this issue will mean looking into how topics pages generate which tables have collapsed versions.

**Comment by tuchandra** [comment]
Race and Hispanic Origin lists `B12002` as having a collapsed version. It therefore appears that all four of these are instances of topics pages incorrectly listing that a collapsed version of a table exists.

**Comment by tuchandra** [comment]
Spent the morning trying to modify the scraper again to look for the special annotations -- `‡§ª†` -- that denote things like iterations, collapsed table, etc. Dealing with Unicode in Python 2 is a headache, though, so I'm looking around for better alternatives (or I may just check the tables manually, who knows).

**Comment by iandees** [comment]
If you can point to a bit of code that's not working, I'm happy to help with Unicode stuff in Python.

**Comment by tuchandra** [comment]
Thanks, Ian. 

I'm using regex to match table codes on the topic pages. `[BC]\d{5}[A-H]?P?R?` has been working fine to capture all of the table codes on a page. Now, some of the pages (e.g., [Children](http://censusreporter.org/topics/children/)) have special characters next to the tables denoting iterations, collapsed tables, etc., like `B09008‡` or `B10051†`. The four characters used are `‡ § ª †`.

Ultimately, I'll want the code to behave differently based on which annotations are there. For right now, let's say that all I want to do is check if the character after the table code is a double-dagger `‡`. I can get the character after with `match.string[match.end() + 1]`, but a comparison always returns False.

```
# -*- coding: utf-8 -*-

...

exp = '[BC]\d{5}[A-H]?P?R?'
all_tables = re.finditer(exp, text)

tables = []
for match in all_tables:
   print match.string[match.start() : match.end() + 1]
   print '‡'
   print match.string[match.end() + 1]
   print (match.string[match.end() + 1] == '‡')
```

will, when I reach the table `B09008` (followed by the double dagger character), print:

```
B09008�
‡
�
False
```

I don't actually care about printing the character properly, but how can I detect and do comparisons with them?

**Comment by iandees** [comment]
First of all, make sure that your terminal is speaking utf8 encoding:

```bash
➜  ~ echo $LANG
en_US.UTF-8
➜  ~ echo $LC_CTYPE
en_US.UTF-8
```

Next, make sure Python is picking up the correct character encoding for stdout:

```python
Python 2.7.11 (default, Dec  5 2015, 14:44:47)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.stdout.encoding
'UTF-8'
```

If you're using Python 2, you should also make sure that the data you're getting in `text` is a Python `unicode` string, not a `str`.

With all those things being assumed, your example seems to work for the most part for me:

```python
>>> import re
>>> exp = '[BC]\d{5}[A-H]?P?R?'
>>> text = u'B09008‡\nB09008\n'
>>> all_tables = re.finditer(exp, text)
>>>
>>> for match in all_tables:
...   print match.string[match.start() : match.end() + 1]
...
B09008‡
B09008
```

The most important change I made here was to tell Python that the strings were `unicode`s, not `str`s (by putting the little `u` in front of the string literals). This is one of the trickiest parts of Python, so I suggest that you watch [this great PyCon US talk](https://www.youtube.com/watch?v=sgHbC6udIqc) that does a pretty good job of explaining it.

Another thing to point out is that you probably should be using regex's matching groups to extract the interesting tidbits instead of using `match.string[start():end()]`. Throw parens around the parts of your regex that you want to extract and then use `match.groups()` to get easy access to the parts of the string without having to do brittle indexing into the matched string:

```python
>>> text = u'Here is table B09008‡ over here, but then B13001 popped way down here!'
>>> exp = '([BC]\d{5}[A-H]?P?R?)(.?) '
>>> all_tables = re.finditer(exp, text)
>>> for match in all_tables:
...   print match.groups()
...
(u'B09008', u'\u2021')
(u'B13001', u'')
```

Notice that the first result there includes a `u'\u2021'` in the second matching group. That's the unicode character ‡ displayed as its hexadecimal equivalent. If you have your output character encoding set up correctly, printing that should give you the expected result:

```python
>>> print u'\u2021'
‡
```

Let me know if you have any questions.

**Comment by tuchandra** [comment]
Thanks, Ian. I finally got it. I did everything you said, and the last step was making sure that I was actually dealing with Unicode strings. The text was obtained from an [HTMLParser](https://docs.python.org/2.7/library/htmlparser.html), and so I had to ensure it was being fed Unicode. I did this with `parser.feed(text.decode('utf-8'))`, and this let me do all of the printing and comparisons I wanted.

**Comment by tuchandra** [comment]
I updated and ran the scraper to test for C tables when the `‡` annotation is listed (and the other annotations, but that's not relevant to this issue). The following tables are listed as having a collapsed version, but there is not one in `census_tabulation_metadata`.

```
Getting Started: []
Age and Sex: []
Children: [u'B22002', u'B09016', u'B05009']
Commute: [u'B08105', u'B08303', u'B08302', u'B08137', u'B08412', u'B08007', u'B08505', u'B08537', u'B08017', u'B08012', u'B08011', u'B08018', u'B08602', u'B08603']
Employment: [u'B23023']
Families: [u'B22007']
Geography: []
Health Insurance: []
Income: []
Migration: [u'B07408', u'B07203', u'B07202']
Poverty: []
Public Assistance: [u'B22007', u'B22002']
Race and Hispanic Origin: [u'B11002', u'B19037', u'B11001', u'B08105', u'B21001', u'B01001', u'B12002', u'B08505', u'B23002', u'B19001', u'B17020', u'B14007', u'B02006', u'B02007', u'B02005', u'B19101', u'B20005']
Same-Sex Couples: []
Seniors: []
Table Codes: []
Veterans and Military: []
```

I'm going to manually confirm these don't exist, then update the branch to fix them on the topic pages.

**Comment by tuchandra** [comment]
And yup -- checked most of them and they don't exist, despite the topic page claiming they do

**Comment by tuchandra** [comment]
See commit 323a1da

**Comment by tuchandra** [comment]
Fixed by #186
